sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageStreamSource.java

```diff
@@ -17,161 +17,565 @@
-    this.parseFn = checkNotNull(parseFn, "parseFn");
-    this.coder = checkNotNull(coder, "coder");
+      Coder<T> outputCoder,
+      BigQueryServices bqServices) {
     this.readSession = checkNotNull(readSession, "readSession");
-    this.streamPosition = checkNotNull(streamPosition, "streamPosition");
-    this.readSizeBytes = readSizeBytes;
+    this.stream = checkNotNull(stream, "stream");
+    this.startOffset = startOffset;
+    this.stopOffset = stopOffset;
+    this.splitDisposition = checkNotNull(splitDisposition, "splitDisposition");
+    this.splitPosition = splitPosition;
+    this.parseFn = checkNotNull(parseFn, "parseFn");
+    this.outputCoder = checkNotNull(outputCoder, "outputCoder");
+    this.bqServices = checkNotNull(bqServices, "bqServices");
+  }
+
+  @VisibleForTesting
+  ReadSession getReadSession() {
+    return readSession;
+  }
+
+  @VisibleForTesting
+  Stream getStream() {
+    return stream;
+  }
+
+  @VisibleForTesting
+  long getStartOffset() {
+    return startOffset;
+  }
+
+  @VisibleForTesting
+  long getStopOffset() {
+    return stopOffset;
+  }
+
+  @VisibleForTesting
+  SplitDisposition getSplitDisposition() {
+    return splitDisposition;
+  }
+
+  @VisibleForTesting
+  StreamPosition getSplitPosition() {
+    return splitPosition;
+  }
+
+  @Override
+  public void validate() {
+    checkState(startOffset >= 0, "startOffset [%s] must be non-negative", startOffset);
+    checkState(stopOffset >= 0, "stopOffset [%s] must be non-negative", stopOffset);
+    checkState(
+        stopOffset >= startOffset,
+        "stopOffset [%s] cannot be less than startOffset [%s]",
+        stopOffset,
+        startOffset);
+    if (splitDisposition == SplitDisposition.SELF) {
+      checkState(splitPosition == null, "Split position cannot be specified with disposition SELF");
+    } else {
+      checkState(
+          splitPosition != null, "Split position must be specified when disposition is not SELF");
+    }
+  }
+
+  @Override
+  public Coder<T> getOutputCoder() {
+    return outputCoder;
+  }
+
+  @Override
+  public void populateDisplayData(DisplayData.Builder builder) {
+    super.populateDisplayData(builder);
+    builder.add(DisplayData.item("readSession", readSession.getName()).withLabel("Read session"));
+    builder.add(DisplayData.item("stream", stream.getName()).withLabel("Stream"));
   }
 
   @Override
-  public List<? extends BoundedSource<T>> split(
-      long desiredBundleSizeBytes, PipelineOptions options) {
-    // Stream source objects can't (currently) be split.
+  public List<? extends BigQueryStorageStreamSource<T>> split(
+      long desiredBundleSizeBytes, PipelineOptions options) throws Exception {
     return ImmutableList.of(this);
   }
 
   @Override
-  public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) {
-    return readSizeBytes != null ? readSizeBytes : 0;
+  public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+    return 0L;
```

Comment:
I don't think there is a good way to provide an estimate of the size of a stream since it can change significantly depending on how much data is read in the context of the underlying stream and in other streams in the read session. It can be anywhere from zero rows to the entire contents of the table.